### ISSUE TITLE
Feature Request: Allow mixed type operations

### ISSUE THREAD

**Issue by jeremysalwen** [viennacl/viennacl-dev#124]
For example, I have an array of floats and a boolean mask

    viennacl::vector<float> a;
    viennacl::vector<bool> b;

I would like to be able to do

    viennacl::vector<float> c=viennacl::linalg::element_prod(a,b)

Right now I just implmeent b as a vector of floats instead.  I'm not sure if this would actually be a performance improvement to use integer types instead, but it would be nice to be able to compare.types.

**Comment by karlrupp** [comment]
Thanks! This request is an extension of https://github.com/viennacl/viennacl-dev/issues/80

Do you have a case where mixed type operations constitute a performance bottleneck? We have not yet run into such a case, hence the priority for mixed type operations remained low.

**Comment by jeremysalwen** [comment]
As I am writing the implementation using ViennaCL, I don't have a way to directly check. (I don't think I could create a well-written custom kernel either.)  I can give you the circumstance in which I thought it seeme natural to perform this operation.

I am calculating on a neural network with dropout.  Without dropout, the calculation of a single layer looks like
({B, A, R} matrices, {x, y, r} vectors)
    B= tanh(A*x+y)

with dropout, we randomly set half of the values of A and y to zero, so the calculation looks like

    B=tanh((A.*R)x+y.*r)

where R is a matrix of random booleans, and r is a vector of random booleans.  The calculation of A.*R in particular is large.

**Comment by karlrupp** [comment]
Considering A.*R in double precision, this would allow for 17 bytes per entry rather than 24 bytes per entry, saving about 30 percent if memory channels can efficiently transfer single bytes. The latter isn't always the case, plus we would have to run an additional just-in-time compilation step with an overhead of about 1 second. Thus, you may still be better off just using the same types ;-)

**Comment by jeremysalwen** [comment]
This operation is far and above the bottleneck of my application, so I don't think the overhead of 1 second will matter.  The 30% memory reduction sounds very promising to me, but I know little to nothing about memory channel efficiency, so I will shut my mouth :)

**Comment by PhilippeTillet** [comment]
Yes, to add to what Karl said, I think it would be certainly possible to
achieve peak bandwidth even for mixed precision operations, but this would
require quite some tuning.
In your drop-out case, I think that you could update your weights matrix
right after training (since doing dropout multiple times in a row is
statistically meaningless):

// post processing

A = element_prod(A, R)

y = element_prod(y, r)

/.../

B=tanh(Ax+y)


You may be able to use R only as a temporary matrix, which will reduce the
memory footprint (hence the PCI-E load) of your program

**Comment by jeremysalwen** [comment]
Hmm, I think things are perhaps a bit unclear.  The main loop of my application is roughly

```
for x in samples
    copy_to_gpu(x)
    R=random_boolean_matrix()
    r=random_boolean_vector()
    B=tanh((A.*R)x+y.*r)
    ...
    A+=some_correction(A,R,x,y,r)
````
```

**Comment by PhilippeTillet** [comment]
Yes, sorry I got confused on drop-out, you are right. In this case, you are
bound by A.*R and y.*r because you are not using mini-batches. GPUs are
very good at matrix-matrix multiplications, but in the matrix-vector case
the PCI-E and OpenCL setup latency is likely to kill you, I would advise
you to preprocess your dataset into an std::list<viennacl::matix> (or if
you use vector, be very, very, very careful about viennacl objects copy
constructors which are deep), so that memory-bound operations don't slow
down your code too much.

**Comment by jeremysalwen** [comment]
Judging based on the metrics here, it seems using integer representations really does improve performance: 

http://cs.nyu.edu/~wanli/dropc/
They get a >3x speed from switching from float to bit representations of the mask

I don't think memeory transfer would actually be an issue if everything is kept on the GPU between iterations (including generating R and r) except the sample x.

**Comment by PhilippeTillet** [comment]
They probably obtain a speed-up because they are using texture memory,
which we don't handle in ViennaCL. It is also hard to say whether they get
a speedup because the kernel is not optimized for global memory, or whether
bitwise representation genuinely helps
